mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala

```diff
@@ -0,0 +1,235 @@
+}
+
+/**
+ * Base representing algorithm for isotonic regression
+ */
+trait IsotonicRegressionAlgorithm
+  extends Serializable {
+
+  /**
+   * Creates isotonic regression model with given parameters
+   *
+   * @param predictions labels estimated using isotonic regression algorithm.
+   *                    Used for predictions on new data points.
+   * @param isotonic isotonic (increasing) or antitonic (decreasing) sequence
+   * @return isotonic regression model
+   */
+  protected def createModel(
+      predictions: Seq[(Double, Double, Double)],
+      isotonic: Boolean): IsotonicRegressionModel
+
+  /**
+   * Run algorithm to obtain isotonic regression model
+   *
+   * @param input (label, feature, weight)
+   * @param isotonic isotonic (increasing) or antitonic (decreasing) sequence
+   * @return isotonic regression model
+   */
+  def run(
+      input: RDD[(Double, Double, Double)],
+      isotonic: Boolean): IsotonicRegressionModel
+}
+
+/**
+ * Parallel pool adjacent violators algorithm for monotone regression
+ */
+class PoolAdjacentViolators private [mllib]
+  extends IsotonicRegressionAlgorithm {
+
+  override def run(
+      input: RDD[(Double, Double, Double)],
+      isotonic: Boolean = true): IsotonicRegressionModel = {
+    createModel(
+      parallelPoolAdjacentViolators(input, isotonic),
+      isotonic)
+  }
+
+  override protected def createModel(
+      predictions: Seq[(Double, Double, Double)],
+      isotonic: Boolean): IsotonicRegressionModel = {
+    new IsotonicRegressionModel(predictions, isotonic)
+  }
+
+  /**
+   * Performs a pool adjacent violators algorithm (PAVA)
+   * Uses approach with single processing of data where violators in previously processed
+   * data created by pooling are fixed immediatelly.
+   * Uses optimization of discovering monotonicity violating sequences
+   * Method in situ mutates input array
+   *
+   * @param in input data
+   * @param isotonic asc or desc
+   * @return result
+   */
+  private def poolAdjacentViolators(
+      in: Array[(Double, Double, Double)],
+      isotonic: Boolean): Array[(Double, Double, Double)] = {
+
+    // Pools sub array within given bounds assigning weighted average value to all elements
+    def pool(in: Array[(Double, Double, Double)], start: Int, end: Int): Unit = {
+      val poolSubArray = in.slice(start, end + 1)
+
+      val weightedSum = poolSubArray.map(lp => lp._1 * lp._3).sum
+      val weight = poolSubArray.map(_._3).sum
+
+      for(i <- start to end) {
+        in(i) = (weightedSum / weight, in(i)._2, in(i)._3)
+      }
+    }
+
+    val isotonicConstraint: (Double, Double) => Boolean = (x, y) => x <= y
+    val antitonicConstraint: (Double, Double) => Boolean = (x, y) => x >= y
+
+    def monotonicityConstraint(isotonic: Boolean) =
+      if(isotonic) isotonicConstraint else antitonicConstraint
+
+    val monotonicityConstraintHolds = monotonicityConstraint(isotonic)
```

Comment:
Of course. I was trying to be as explicit as possible for reader's clarity. Happy to change that.